powerauth-fido2-tests/src/main/java/com/wultra/security/powerauth/fido2/service/AssertionService.java

```diff
@@ -0,0 +1,123 @@
+import com.wultra.security.powerauth.client.model.request.fido2.AssertionChallengeRequest;
+import com.wultra.security.powerauth.client.model.request.fido2.AssertionVerificationRequest;
+import com.wultra.security.powerauth.client.model.response.fido2.AssertionChallengeResponse;
+import com.wultra.security.powerauth.client.model.response.fido2.AssertionVerificationResponse;
+import com.wultra.security.powerauth.fido2.configuration.WebAuthnConfiguration;
+import com.wultra.security.powerauth.fido2.controller.request.AssertionOptionsRequest;
+import com.wultra.security.powerauth.fido2.controller.request.VerifyAssertionRequest;
+import com.wultra.security.powerauth.fido2.controller.response.AssertionOptionsResponse;
+import com.wultra.security.powerauth.fido2.controller.response.CredentialDescriptor;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StringUtils;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Service for WebAuthn authentication tasks.
+ *
+ * @author Jan Pesek, jan.pesek@wultra.com
+ */
+@Service
+@AllArgsConstructor
+@Slf4j
+public class AssertionService {
+
+    private static final String OPERATION_DATA_EXTENSION_KEY = "txAuthSimple";
+
+    private final PowerAuthFido2Client fido2Client;
+    private final SharedService sharedService;
+    private final WebAuthnConfiguration webauthnConfig;
+
+    /**
+     * Build public key assertion options.
+     * @param request Request form with user input.
+     * @return Public key assertion options.
+     * @throws PowerAuthClientException in case of PowerAuth server communication error.
+     */
+    public AssertionOptionsResponse assertionOptions(final AssertionOptionsRequest request) throws PowerAuthClientException {
+        final String userId = request.username();
+        final String applicationId = request.applicationId();
+
+        logger.info("Building assertion options for userId={}, applicationId={}", userId, applicationId);
+
+        final List<CredentialDescriptor> existingCredentials = sharedService.fetchExistingCredentials(userId, applicationId);
+        if (existingCredentials.isEmpty() && StringUtils.hasText(userId))  {
+            throw new IllegalStateException("Not registered yet.");
+        }
+
+        final AssertionChallengeResponse challengeResponse = fetchChallenge(applicationId, request.templateName());
+        final String challenge = challengeResponse.getChallenge();
+        final String operationData = extractOperationData(challenge);
+
+        return AssertionOptionsResponse.builder()
+                .challenge(challenge)
+                .rpId(webauthnConfig.getRpId())
+                .timeout(webauthnConfig.getTimeout().toMillis())
+                .allowCredentials(existingCredentials)
+                .userVerification(UserVerificationRequirement.PREFERRED)
```

Comment:
I believe using `UserVerificationRequirement.PREFERRED` is a bad practice, because it allows amiguity. Can we use the `REQUIRED` option? I believe we should not accept authenticators which do not ahve the user verification option.

Reply:
I removed the user verification settings from backend. It will be configurable via UI with `required` as default

<img src="https://github.com/wultra/powerauth-tests/assets/66425344/a3ac2b28-d00c-4d65-9693-08c819a7463e" width="200">